### DISCUSSION TITLE
invalid mods per rulesets osu lazer

### DISCUSSION THREAD

**Discussion by goldengun111** [ppy/osu#17305] [Q&A]
On lazer multiplayer its not letting me choose any multiplayer mods be it one mod or all mods available 
no clue what the AS ruleset refers to 

Lazer version: 2022.314.0-lazer (latest)
Windows version: win 10 10.0.19043 build 19043

https://imgur.com/0DZGUze
[database.log](https://github.com/ppy/osu/files/8289301/database.log)
[legacy-ipc.log](https://github.com/ppy/osu/files/8289302/legacy-ipc.log)
[performance.log](https://github.com/ppy/osu/files/8289304/performance.log)
[runtime.log](https://github.com/ppy/osu/files/8289305/runtime.log)
[updater.log](https://github.com/ppy/osu/files/8289306/updater.log)

**Comment by bdach** [comment]
This error is originating from the spectator server: https://github.com/ppy/osu-server-spectator/blob/08d785461d57cc56ed42abc3539307148619e76b/osu.Server.Spectator/Extensions/MultiplayerPlaylistItemExtensions.cs#L63-L74

@peppy what's the production instance of spectator server at? Is it using latest game nuget? If not, then the helper method that generates the list of valid mods for ruleset won't return the new adaptive speed mod.

That said it's about to get deleted from multiplayer anyway (https://github.com/ppy/osu/pull/17270) so maybe not worth fixing...

**Reply by peppy** [reply]
There definitely wasn't a new version deployed, but I can't understand this user not being able to set *any* mod.

I've tested and non-AS mods work fine here.

**Reply by goldengun111** [reply]
I've tried just difficulty adjust by itself as a free mod and it wouldn't work, then I've tried the entire conversion category as free mods and it wouldn't work either and selecting all of the mods available still gave me a AS ruleset error

**Reply by bdach** [reply]
i think the user may be trying to enable _all_ free mods but failing because of AS being selected? hard to tell.

**Reply by bdach** [reply]
@goldengun111 that is definitely not happening here. difficulty adjust looks to be working just fine. see screenshot below from created room.

![osu_2022-03-17_20-36-10](https://user-images.githubusercontent.com/20418176/158881950-62b287d6-c645-4243-8032-3a94694ffc36.jpg)

**Reply by goldengun111** [reply]
Seems like adaptive speed is the issue, like the abbreviation, didn't realise, it seems like just not choosing that fixes it, thanks

**Comment by hlysine** [comment]
Not sure if this is a cause for concern, but it seems like the mod validation logic is different when a multi room is first created.

If you choose AS mod via the "Add item" button when the room is already created, it throws the error `Invalid mods were selected for specified ruleset: AS`. But if you choose AS mod via the "Select beatmap" button in the room creation screen, the mod gets enabled just fine and you can even start a game with the mod on.

**Reply by bdach** [reply]
initial validation is done by web, and subsequent checks are made by spectator server *and* web. it's a known idiosyncrasy - the validation logic would either have to match exactly in both places, or the spectator server would have to middleman in the room creation process.
